### PR TITLE
Parsing JSON Notices

### PR DESCRIPTION
**Pull request by Vidushi-GitHub**
# Description

This document includes how to parse, encode, decode the JSON Schema files.

Resolves: #2149

### CODE REVIEW

**Comment by Vidushi-GitHub** [comment]
Screenshot 2024-08-02 at 2 54 40 PM

File: app/routes/docs.client.samples.md [hunk 1]
```diff
@@ -164,3 +166,62 @@ for message in consumer.consume(end[0].offset - start[0].offset, timeout=1):
             continue
     print(message.value())
 ```
+
+# Working With JSON Schema
```
Comment (Vidushi-GitHub) [inline app/routes/docs.client.samples.md:170]:
```suggestion
## Working With JSON Schema
```

File: app/routes/docs.client.samples.md [hunk 2]
```diff
@@ -164,3 +166,62 @@ for message in consumer.consume(end[0].offset - start[0].offset, timeout=1):
...
+with open('sample.example.json', 'r') as example_file:
+    example = json.load(example_file)
+
+print('Schema:', schema)
+print('Example:', example)
+```
+
+## Encoding and Decoding of Embedded Data
+
+The following code sample shows how to encode/decode a file in Python. The `base64` package includes the methods `b64decode` and `b64encode` to make this task simple.
```
Comment (tylerbarna) [inline app/routes/docs.client.samples.md:194]:
```suggestion
The following code shows how to encode/decode a GCN Notice using the `b64decode` and `b64encode` methods from the `base64` Python package.
```

File: app/routes/docs.client.samples.md [hunk 3]
```diff
@@ -164,3 +166,62 @@ for message in consumer.consume(end[0].offset - start[0].offset, timeout=1):
...
+print(encoded_string)
+# b'a1512dabc1b6adb3cd1b6dcb6d4c6......'
+
+# Decode and write the content to a local file:
+with open("path/to/destination/file", 'wb') as file:
+    file.write(base64.b64decode(encoded_string))
+
+```
+
+For example, if we wanted to include a FITS file in a notice, you would need to add a property to your schema definition and define the value to be the following:
```
Comment (tylerbarna) [inline app/routes/docs.client.samples.md:212]:
```suggestion
If you want to include a FITS file in a Notice, you add a property to your schema definition in the following format:
```

File: app/routes/docs.client.samples.md [hunk 4]
```diff
@@ -164,3 +166,62 @@ for message in consumer.consume(end[0].offset - start[0].offset, timeout=1):
...
+type: 'string',
+contentEncoding: 'base64',
+contentMediaType: 'image/fits',
+},
+null,
+2
+)}
+/>
+
+Then in your data production pipeline, you can use the encoding steps to convert your file to a bytestring, and set the value of the property to said bytestring. See [non JSON data](https://json-schema.org/understanding-json-schema/reference/non_json_data.html) for more information.
```
Comment (tylerbarna) [inline app/routes/docs.client.samples.md:227]:
```suggestion
In your data production pipeline, you can use the encoding steps to convert your file to a bytestring and set the value of the property to this bytestring. See [non-JSON data](https://json-schema.org/understanding-json-schema/reference/non_json_data.html) for more information.
```

**Review by tylerbarna — CHANGES_REQUESTED** [review]

**Comment by dakota002** [comment]
@Vidushi-GitHub Is this PR ready for review or still a draft?

**Comment by Vidushi-GitHub** [comment]
> @Vidushi-GitHub Is this PR ready for review or still a draft?

Ready to review. 

> @Vidushi-GitHub Is this PR ready for review or still a draft?

Ready to review, moved, thanks.

File: app/routes/docs.client.samples.md (hunk 2 above)
Comment (Vidushi-GitHub) [reply app/routes/docs.client.samples.md:194]: I still find the previous more descriptive.

File: app/routes/docs.client.samples.md [hunk 5]
```diff
@@ -164,3 +166,62 @@ for message in consumer.consume(end[0].offset - start[0].offset, timeout=1):
...
+{
+type: 'string',
+contentEncoding: 'base64',
+contentMediaType: 'image/fits',
+},
+null,
+2
+)}
+/>
+
```
Comment (dakota002) [inline app/routes/docs.client.samples.md:226]: @Vidushi-GitHub can you attach a screenshot of this code sample? Does it render correctly in the browser?

File: app/routes/docs.client.samples.md [hunk 6]
```diff
@@ -164,3 +166,62 @@ for message in consumer.consume(end[0].offset - start[0].offset, timeout=1):
...
+code={JSON.stringify(
+{
+type: 'string',
+contentEncoding: 'base64',
+contentMediaType: 'image/fits',
+},
+null,
+2
+)}
+/>
```
Comment (lpsinger) [inline app/routes/docs.client.samples.md:225]: Please fence code using triple backticks, ```, not the Highlight component. We only use the Highlight component when the code sample is generated programmatically.

Also, this does not look like valid JSON code.

File: app/routes/docs.client.samples.md [hunk 7]
```diff
@@ -164,3 +166,62 @@ for message in consumer.consume(end[0].offset - start[0].offset, timeout=1):
             continue
     print(message.value())
 ```
+
+## Working With JSON Schema
+
+GCN Notices are distributed in JSON format. This guide describes how to programmatically read JSON schema.
```
Comment (lpsinger) [inline app/routes/docs.client.samples.md:172]: Some GCN Notices are distributed in JSON format, but not all..

File: app/routes/docs.client.samples.md [hunk 8]
```diff
@@ -164,3 +166,62 @@ for message in consumer.consume(end[0].offset - start[0].offset, timeout=1):
             continue
     print(message.value())
 ```
+
+## Working With JSON Schema
```
Comment (lpsinger) [inline app/routes/docs.client.samples.md:170]: Please rename this section and the previous one to make it clear that one section is dealing with decoding XML data and the other is dealing with JSON data.

File: app/routes/docs.client.samples.md [hunk 9]
```diff
@@ -164,3 +166,62 @@ for message in consumer.consume(end[0].offset - start[0].offset, timeout=1):
...
+# Load and parse schema and example JSON files
+with open('sample.schema.json', 'r') as schema_file:
+    schema = json.load(schema_file)
+
+with open('sample.example.json', 'r') as example_file:
+    example = json.load(example_file)
+
+print('Schema:', schema)
+print('Example:', example)
+```
```
Comment (lpsinger) [inline app/routes/docs.client.samples.md:190]: Please provide sample code for parsing Kafka records in a message consuming loop.

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/routes/docs.client.samples.md (hunk 5 above)
Comment (Vidushi-GitHub) [reply app/routes/docs.client.samples.md:226]: It was attached above too. 
Screenshot 2024-08-06 at 3 37 44 PM

File: app/routes/docs.client.samples.md (hunk 7 above)
Comment (Vidushi-GitHub) [inline app/routes/docs.client.samples.md:172]:
```suggestion
For new missions, GCN Notices are preferably distributed in JSON format. This guide describes how to programmatically read the JSON schema.
```

Comment (lpsinger) [reply app/routes/docs.client.samples.md:172]: Why are you making suggestions on your own PR? Just make the change.

File: app/routes/docs.client.samples.md (hunk 6 above)
Comment (Vidushi-GitHub) [reply app/routes/docs.client.samples.md:225]: Please see: https://json-schema.org/understanding-json-schema/reference/non_json_data
for valid JSON code

File: app/routes/docs.client.samples.md [hunk 10]
```diff
@@ -164,3 +164,57 @@ for message in consumer.consume(end[0].offset - start[0].offset, timeout=1):
...
     print(message.value())
 ```
+
+## Working With JSON Schema
+
+For new missions, GCN Notices are preferably distributed in JSON format. This guide describes how to programmatically read the JSON schema.
+
+## Parsing JSON
+
+Read the JSON data from [sample.schema.json](https://gcn.nasa.gov/docs/notices/schema) and [sample.example.json](https://gcn.nasa.gov/docs/notices/schema), which parses it into Python dictionaries.
```
Comment (lpsinger) [inline app/routes/docs.client.samples.md:174]: Neither of these links are to JSON schema documents that exist.

File: app/routes/docs.client.samples.md (hunk 2 above)
Comment (lpsinger) [reply app/routes/docs.client.samples.md:192]: Provide hyperlinks to these Python methods.

File: app/routes/docs.client.samples.md [hunk 11]
```diff
@@ -164,3 +164,57 @@ for message in consumer.consume(end[0].offset - start[0].offset, timeout=1):
...
+
+If you want to include a FITS file in a Notice, you add a property to your schema definition in the following format:
+
+```python
+{
+    type: 'string',
+    contentEncoding: 'base64',
+    contentMediaType: 'image/fits',
+}
+```
```
Comment (lpsinger) [inline app/routes/docs.client.samples.md:218]: This should be formatted as JSON, not as Python code.

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Comment by bob-wiegand** [comment]
Is there a standard approach for conveying a (suggested) filename in the notice?
Is a key other than contentMediaType in the schema going to convey compression if present? For example, can the schema convey that a value is a (base64 encoded) gzipped PDF?

**Comment by Vidushi-GitHub** [comment]
A new PR (https://github.com/nasa-gcn/gcn.nasa.gov/pull/2669) is created, 
as this one had lots of files touched in rebase process.

**Pull request closed without merging**